src/shared/pam-util.c

```diff
@@ -50,6 +50,18 @@ int pam_syslog_pam_error(pam_handle_t *handle, int level, int error, const char
 }
 
 static void cleanup_system_bus(pam_handle_t *handle, void *data, int error_status) {
+        /* The PAM_DATA_SILENT flag is the way that pam_end() communicates to the module stack that this
+         * invocation of pam_end() is not the final one, but in the process that is going to directly exec
+         * the child. This means we are being called after a fork(), and we do not want to try and clean
+         * up the sd-bus object, as the file descriptors (which are CLOEXEC) will be invalid, and we'll hit
+         * an assertion. */
+        if (error_status & PAM_DATA_SILENT)
```

Comment:
while i know that this pam guy claims PAM_DATA_SILENT has this meaning i can't find this documented, and even if it was supposed to be used this way, I doubt this will be a universally accepted mechanism.

Hence I think we should rather check pids explicitly: store away inside the PAM handle which PID we allocated the bus in, and only destroy it if we are still in the same pid.

Reply:
actually it is documented, but in a different manpage: https://linux.die.net/man/3/pam_end

Reply:
> This argument can be logically OR'd with PAM_DATA_SILENT to indicate to indicate that the module should not treat the call too seriously. It is generally used to indicate that the current closing of the library is in a [fork](https://linux.die.net/man/2/fork)(2)ed process, and that the parent will take care of cleaning up things that exist outside of the current process space (files etc.).

Reply:
I've opened a PR to document that in the callback description too so that it's immediately visible when checking how that works: https://github.com/linux-pam/linux-pam/pull/566

Reply:
I'd still prefer if we looked at the actual condition here, i.e. at the pid, not the flag. 

so what about this: we do what we do with sd-bus and attach a PID to pam_handle_t. i.e. with some helper func we'll store the pid in a data field, and then we check it again whenever needed. i.e.:

```c
int pam_validate_pid(pam_handle_t h) {
        const void *p;
        int r;

        /* If this function wasn't called on the handle yet, installs the current PID on it. Otherwise checks if the installed PID matches our own, and if not returns PAM_BAD_ITEM. */

        r = pam_get_data(handle, "systemd-pid", &p);
        if (r == PAM_NO_MODULE_DATA) 
                return pam_det_data(handle, "systemd-pid", PID_TO_PTR(getpid_cached()), NULL);
        if (r != PAM_SUCCESS) 
                return r;

        return PTR_TO_PID(p) == getpid_cached() ? PAM_SUCCESS : PAM_BAD_ITEM;
}
```

And then in all our pam hooks, just add first thing:

```c
r = pam_validate_pid(h);
if (r != PAM_SUCCESS)
        return r;
```

Reply:
More cases like that where cleanup should be skipped could be added to pam, that are not necessarily PID related. Given that's the documented interface, we should use it.
The PID stuff we can do in sd-bus itself. I do not want to go overboard here, because this fixes a bug and I want to backport it to stable. I'll do the PID stuff in a different PR, that will not be backported.

Reply:
> More cases like that where cleanup should be skipped could be added to pam, that are not necessarily PID related.

Yeah, but that's precisely the problem: we only care for the fork()/pid change case. We don#t care for other reasons. This is about closing of bus connections, nothing else. And for that only the fork()/pid thing matters, nothing else.

> Given that's the documented interface, we should use it.

Well, it's documented at the one place noone looks at and at the places people actually look (you and me included) it isn't, but suggests it is about log messages, nothing else.

And even if you look at the docs, this is what it says:

> This argument can be logically OR'd with PAM_DATA_SILENT to indicate that the module should not take the call too seriously. It is generally used to indicate that the current closing of the library is in a fork(2)ed process, and that the parent will take care of cleaning up things that exist outside of the current process space (files etc.).

This says the flag is generally set in case of fork(), but this really doesn't say *only* in that case. Instead it just says something about "not taking it seriously", for whatever this means.

What I am more concerned about is that noone actually sets the flag correctly IRL. We didn't either for PAMName= until last year! (7feb2b5737a)

Hence: sorry, we shouldn#t rely on the flag. I think we should strive for setting it as it apparently is intended to be used when we create a PAM session, but I am very sure we should not rely on it when we are loaded as part of a foreign PAM sessions. In particular as it doesn't really indicate striuctly what we want to check for...